scripts/pylib/twister/twisterlib/handlers.py

```diff
@@ -493,7 +497,9 @@ def make_dut_available(self, dut):
             dut.failures_increment()
         logger.debug(f"Release DUT:{dut.platform}, Id:{dut.id}, "
                      f"counter:{dut.counter}, failures:{dut.failures}")
-        dut.available = 1
+        for _d in self.duts:
+            if _d.id == dut.id:
+                _d.available = 1
```

Comment:
I am worried this is theoretically problematic. I think there is some very unlikely race condition. Since there is no additional lock here, it can happen that some duts with the same id will be already available, whereas the rest still not, but another thread will try to use the just-released instance hence locking all duts with the same id. Not sure if it makes sense... I guess locking all duts with the same id here before making them might prevent this issue

Reply:
Good catch! Thanks!
I've added locking of all shared DUTs here as well.
To avoid repeating the same code, I've extracted the locking part into the `acquire_dut_locks` method, which uses the `contextmanager` decorator to allow using the `with ...:` construction.